### ISSUE TITLE
The gradient of projection haven't been detached in this implementation of SimSiam.

### ISSUE THREAD

**Issue by chenrz925** [lightly-ai/lightly#559]
The projection `z0` and `z1` of raw output of backbone in SimSiam should be detached. 

As the version of [facebookresearch/simsiam](https://github.com/facebookresearch/simsiam/blob/main/simsiam/builder.py), the projection's gradients have been detached finally: 

```python
class SimSiam(nn.Module):
...
    def forward(self, x1, x2):
        """
        Input:
            x1: first views of images
            x2: second views of images
        Output:
            p1, p2, z1, z2: predictors and targets of the network
            See Sec. 3 of https://arxiv.org/abs/2011.10566 for detailed notations
        """

        # compute features for one view
        z1 = self.encoder(x1) # NxC
        z2 = self.encoder(x2) # NxC

        p1 = self.predictor(z1) # NxC
        p2 = self.predictor(z2) # NxC

        return p1, p2, z1.detach(), z2.detach()
```

The implementation of this project: 

```python
class SimSiam(nn.Module):
...
    def forward(self, 
                x0: torch.Tensor, 
                x1: torch.Tensor = None,
                return_features: bool = False):
        f0 = self.backbone(x0).flatten(start_dim=1)
        z0 = self.projection_mlp(f0)
        p0 = self.prediction_mlp(z0)

        out0 = (z0, p0)

        # append features if requested
        if return_features:
            out0 = (out0, f0)

        if x1 is None:
            return out0
        
        f1 = self.backbone(x1).flatten(start_dim=1)
        z1 = self.projection_mlp(f1)
        p1 = self.prediction_mlp(z1)

        out1 = (z1, p1)

        # append features if requested
        if return_features:
            out1 = (out1, f1)

        return out0, out1
```

Would you please check if this is inconsistent with the paper of SimSiam? Thanks a lot.

**Comment by philippmwirth** [comment]
Closed by #503
